File: src/tokens.ts
export enum TokenType {
	// Literals
	STRING,
	NUMBER,
	IDENT,

	// Keywords
	NULL,
	TRUE,
	FALSE,

	// Type keywords
	STRING_TYPE, // "string"
	INT_TYPE, // "int"
	FLOAT_TYPE, // "float"
	BOOL_TYPE, // "bool"
	NUMBER_TYPE, // "number"
	BYTES_TYPE, // "bytes"

	// Punctuation
	LBRACE, // {
	RBRACE, // }
	LBRACKET, // [
	RBRACKET, // ]
	COLON, // :
	COMMA, // ,
	LPAREN, // (
	RPAREN, // )

	// Operators
	PIPE, // |
	AMP, // &
	GTE, // >=
	LTE, // <=
	GT, // >
	LT, // <
	EQ, // ==
	NEQ, // !=
	MATCH, // =~
	NOT_MATCH, // !~

	// Special
	COMMENT, // // ...
	ELLIPSIS, // ...
	QUESTION, // ?
	HASH, // #
	UNDERSCORE, // _ (top)
	BOTTOM, // _|_

	EOF,
}

export interface Token {
	type: TokenType;
	value: string;
	line: number;
	column: number;
}

File: wasm/src/tokens.rs
#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    // Literals
    String,
    Number,
    Ident,

    // Keywords
    Null,
    True,
    False,

    // Type keywords
    StringType,
    IntType,
    FloatType,
    BoolType,
    NumberType,
    BytesType,

    // Punctuation
    LBrace,
    RBrace,
    LBracket,
    RBracket,
    Colon,
    Comma,
    LParen,
    RParen,

    // Operators
    Pipe,
    Amp,
    Gte,
    Lte,
    Gt,
    Lt,
    Eq,
    Neq,
    Match,
    NotMatch,

    // Special
    Comment,
    Ellipsis,
    Question,
    Hash,
    Underscore,
    Bottom,

    Eof,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub value: std::string::String,
    pub line: usize,
    pub column: usize,
}

impl Token {
    pub fn new(token_type: TokenType, value: impl Into<std::string::String>, line: usize, column: usize) -> Self {
        Self {
            token_type,
            value: value.into(),
            line,
            column,
        }
    }
}
